### PR TITLE
Point Set Processing: Callbacks

### PR DESCRIPTION
**Pull request by sgiraudot**
## Summary of Changes

Not-yet-approved [small feature](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Callbacks_for_Point_Set_Processing) that introduces a named parameter `callback` in some functions of PSP.

## Release Management

* Affected package(s): Point Set Processing, Polyhedron demo
* Issue(s) solved (if any): --
* Feature/Small Feature (if any): [Callbacks_for_Point_Set_Processing](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Callbacks_for_Point_Set_Processing)
* pre-approved by Laurent Rineau 11:11, 30 May 2018

### CODE REVIEW

**Comment by lrineau** [comment]
The file `CHANGES.md` should be modified.

**Comment by lrineau** [comment]
I also found:
```
[ 87%] Building CXX object Plugins/Point_set/CMakeFiles/point_set_average_spacing_plugin.dir/Point_set_average_spacing_plugin.cpp.o
cd /home/cgal_tester/build/src/cmake/platforms/CentOS7-Release/test/Polyhedron_Demo/Plugins/Point_set && /usr/bin/c++   -DCGAL_EIGEN3_ENABLED -DCGAL_TEST_SUITE=1 -DCGAL_USE_GMP -DCGAL_USE_MPFR -DCGAL_USE_ZLIB=1 -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NO_DEBUG -DQT_NO_KEYWORDS -DQT_OPENGL_LIB -DQT_SCRIPT_LIB -DQT_SVG_LIB -DQT_WIDGETS_LIB -DQT_XML_LIB -DSCENE_IMAGE_GL_BUFFERS_AVAILABLE -DUSE_FORWARD_DECL -Dpoint_set_average_spacing_plugin_EXPORTS -I/home/cgal_tester/build/src/cmake/platforms/CentOS7-Release/test/Polyhedron_Demo/Plugins/Point_set -I/home/cgal_tester/build/src/cmake/platforms/CentOS7-Release/test/Polyhedron_Demo/./CGAL_demo -I/home/cgal_tester/build/src/cmake/platforms/CentOS7-Release/test/Polyhedron_Demo/../../include -I/home/cgal_tester/build/src/cmake/platforms/CentOS7-Release/test/Polyhedron_Demo/./include -I/home/cgal_tester/build/src/cmake/platforms/CentOS7-Release/test/Polyhedron_Demo/. -I/home/cgal_tester/build/src/cmake/platforms/CentOS7-Release/test/Polyhedron_Demo -I/home/cgal_tester/build/src/cmake/platforms/CentOS7-Release/include -I/mnt/testsuite/include -isystem /usr/include/eigen3 -I/usr/include/qt5 -I/usr/include/qt5/QtOpenGL -I/usr/include/qt5/QtWidgets -I/usr/include/qt5/QtGui -I/usr/include/qt5/QtCore -I/usr/lib64/qt5/./mkspecs/linux-g++ -I/usr/include/qt5/QtSvg -I/usr/include/qt5/QtXml -I/usr/include/qt5/QtScript  -Wall -frounding-math -O3 -DCGAL_NDEBUG -fPIC   -DNOMINMAX -DCGAL_LINKED_WITH_TBB -Wall -frounding-math -fPIC -std=gnu++11 -o CMakeFiles/point_set_average_spacing_plugin.dir/Point_set_average_spacing_plugin.cpp.o -c /home/cgal_tester/build/src/cmake/platforms/CentOS7-Release/test/Polyhedron_Demo/Plugins/Point_set/Point_set_average_spacing_plugin.cpp
In file included from /mnt/testsuite/include/CGAL/compute_average_spacing.h:43:0,
                 from /home/cgal_tester/build/src/cmake/platforms/CentOS7-Release/test/Polyhedron_Demo/Plugins/Point_set/Point_set_average_spacing_plugin.cpp:8:
/mnt/testsuite/include/CGAL/internal/Parallel_callback.h:37:3: error: 'atomic' in namespace 'CGAL::cpp11' does not name a type
   cpp11::atomic<std::size_t>* m_advancement;
   ^
/mnt/testsuite/include/CGAL/internal/Parallel_callback.h:38:3: error: 'atomic' in namespace 'CGAL::cpp11' does not name a type
   cpp11::atomic<bool>* m_interrupted;
   ^
/mnt/testsuite/include/CGAL/internal/Parallel_callback.h:41:3: error: 'thread' in namespace 'CGAL::cpp11' does not name a type
   cpp11::thread* m_thread;
   ^
/mnt/testsuite/include/CGAL/internal/Parallel_callback.h:91:3: error: 'atomic' in namespace 'CGAL::cpp11' does not name a type
   cpp11::atomic<std::size_t>& advancement() { return *m_advancement; }
   ^
/mnt/testsuite/include/CGAL/internal/Parallel_callback.h:92:3: error: 'atomic' in namespace 'CGAL::cpp11' does not name a type
   cpp11::atomic<bool>& interrupted() { return *m_interrupted; }
   ^
```
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.13-Ic-39/Polyhedron_Demo/TestReport_lrineau_CentOS7-Release.gz

And that is strange because the polyhedron demo is compiled with C++11.

**Comment by sgiraudot** [comment]
The Windows errors should be solved.

For the CentOS one, it seems that `CGAL_CFG_NO_THREAD` is set to 1. `gcc` is used in its version 4.8 and, quoting [the source code of Boost](https://www.boost.org/doc/libs/1_67_0/boost/config/stdlib/libstdcpp3.hpp):
```c++
//  C++0x features in GCC 4.8.0 and later
//
#if (BOOST_LIBSTDCXX_VERSION < 40800) || !defined(BOOST_LIBSTDCXX11)
// Note that although <atomic> existed prior to gcc 4.8 it was largely unimplemented for many types:
#  define BOOST_NO_CXX11_HDR_ATOMIC
#  define BOOST_NO_CXX11_HDR_THREAD
#endif
```
So basically, `std::thread` is considered unavailable even if C++11 is activated. I'm not sure what we should do about that… I tried to bypass the test `CGAL_CFG_NO_THREAD` before including `<thread>` and it compiled without complaining (I cannot test the demo as it's on Docker, though).

Weirdly, I cannot reproduce on my OS even using G++4.8…

Maybe we should just assume that, if TBB does not provide its `std::thread` implementation, then it means that the STL one is available, and remove this `CGAL_CFG_NO_THREAD`?

**Comment by lrineau** [comment]
[**@sgiraudot**](https://github.com/sgiraudot) commented on [Jun 18, 2018, 4:27 PM GMT+2]> Maybe we should just assume that, if TBB does not provide its `std::thread` implementation, then it means that the STL one is available, and remove this `CGAL_CFG_NO_THREAD` test?

That is what I understand from https://software.intel.com/en-us/blogs/2010/05/04/transitioning-to-tbb-30

It seems that since TBB 3.0 (May 2010), they have decided that the `thread` API is the one of `std::thread`: either it is available, or TBB provides a fake `std::thread` API.

**Comment by sgiraudot** [comment]
The Point Set Processing line looks good. There is still an error in the demo (almost on every platform) because of multiple definitions of `sleep_for()`, which I just fixed with https://github.com/CGAL/cgal/pull/3129/commits/96dcfc070dbf2f8c33c5f86c557f144eed791559.

**Comment by lrineau** [comment]
There is also a dllimport/dllexport issue, on Windows:
```
[ 33%] Building CXX object CMakeFiles/scene_callback_signaler.dir/Callback_signaler.cpp.obj
	C:\PROGRA~2\MICROS~2.0\VC\bin\amd64\cl.exe @C:\Users\afabri\AppData\Local\Temp\nm7BC0.tmp
Callback_signaler.cpp
C:\CGAL\test\CGAL-4.13-Ic-42\cmake\platforms\x64_Cygwin-Windows10_MSVC2013-Debug-64bits\test\Polyhedron_Demo\Callback_signaler.cpp(4) : warning C4273: 'Callback_signaler::Callback_signaler' : inconsistent dll linkage
        c:\cgal\test\cgal-4.13-ic-42\cmake\platforms\x64_cygwin-windows10_msvc2013-debug-64bits\test\polyhedron_demo\Callback_signaler.h(22) : see previous definition of '{ctor}'
C:\CGAL\test\CGAL-4.13-Ic-42\cmake\platforms\x64_Cygwin-Windows10_MSVC2013-Debug-64bits\test\Polyhedron_Demo\Callback_signaler.cpp(9) : warning C4273: 'Callback_signaler::emit_signal' : inconsistent dll linkage
        c:\cgal\test\cgal-4.13-ic-42\cmake\platforms\x64_cygwin-windows10_msvc2013-debug-64bits\test\polyhedron_demo\Callback_signaler.h(24) : see previous definition of 'emit_signal'
[ 33%] Building CXX object CMakeFiles/scene_callback_signaler.dir/scene_callback_signaler_autogen/mocs_compilation.cpp.obj
	C:\PROGRA~2\MICROS~2.0\VC\bin\amd64\cl.exe @C:\Users\afabri\AppData\Local\Temp\nm7DB5.tmp
mocs_compilation.cpp
c:\cgal\test\cgal-4.13-ic-42\cmake\platforms\x64_cygwin-windows10_msvc2013-debug-64bits\test\polyhedron_demo\scene_callback_signaler_autogen\EWIEGA46WW/moc_Callback_signaler.cpp(74) : warning C4273: 'Callback_signaler::qt_static_metacall' : inconsistent dll linkage
        c:\cgal\test\cgal-4.13-ic-42\cmake\platforms\x64_cygwin-windows10_msvc2013-debug-64bits\test\polyhedron_demo\scene_callback_signaler_autogen\ewiega46ww\../../Callback_signaler.h(16) : see previous definition of 'qt_static_metacall'
c:\cgal\test\cgal-4.13-ic-42\cmake\platforms\x64_cygwin-windows10_msvc2013-debug-64bits\test\polyhedron_demo\scene_callback_signaler_autogen\EWIEGA46WW/moc_Callback_signaler.cpp(96) : warning C4273: 'staticMetaObject' : inconsistent dll linkage
        c:\cgal\test\cgal-4.13-ic-42\cmake\platforms\x64_cygwin-windows10_msvc2013-debug-64bits\test\polyhedron_demo\scene_callback_signaler_autogen\ewiega46ww\../../Callback_signaler.h(16) : see previous definition of 'public: static QMetaObject const Callback_signaler::staticMetaObject'
c:\cgal\test\cgal-4.13-ic-42\cmake\platforms\x64_cygwin-windows10_msvc2013-debug-64bits\test\polyhedron_demo\scene_callback_signaler_autogen\EWIEGA46WW/moc_Callback_signaler.cpp(96) : error C2491: 'Callback_signaler::staticMetaObject' : definition of dllimport static data member not allowed
c:\cgal\test\cgal-4.13-ic-42\cmake\platforms\x64_cygwin-windows10_msvc2013-debug-64bits\test\polyhedron_demo\scene_callback_signaler_autogen\EWIEGA46WW/moc_Callback_signaler.cpp(103) : warning C4273: 'Callback_signaler::metaObject' : inconsistent dll linkage
        c:\cgal\test\cgal-4.13-ic-42\cmake\platforms\x64_cygwin-windows10_msvc2013-debug-64bits\test\polyhedron_demo\scene_callback_signaler_autogen\ewiega46ww\../../Callback_signaler.h(16) : see previous definition of 'metaObject'
c:\cgal\test\cgal-4.13-ic-42\cmake\platforms\x64_cygwin-windows10_msvc2013-debug-64bits\test\polyhedron_demo\scene_callback_signaler_autogen\EWIEGA46WW/moc_Callback_signaler.cpp(108) : warning C4273: 'Callback_signaler::qt_metacast' : inconsistent dll linkage
        c:\cgal\test\cgal-4.13-ic-42\cmake\platforms\x64_cygwin-windows10_msvc2013-debug-64bits\test\polyhedron_demo\scene_callback_signaler_autogen\ewiega46ww\../../Callback_signaler.h(16) : see previous definition of 'qt_metacast'
c:\cgal\test\cgal-4.13-ic-42\cmake\platforms\x64_cygwin-windows10_msvc2013-debug-64bits\test\polyhedron_demo\scene_callback_signaler_autogen\EWIEGA46WW/moc_Callback_signaler.cpp(116) : warning C4273: 'Callback_signaler::qt_metacall' : inconsistent dll linkage
        c:\cgal\test\cgal-4.13-ic-42\cmake\platforms\x64_cygwin-windows10_msvc2013-debug-64bits\test\polyhedron_demo\scene_callback_signaler_autogen\ewiega46ww\../../Callback_signaler.h(16) : see previous definition of 'qt_metacall'
c:\cgal\test\cgal-4.13-ic-42\cmake\platforms\x64_cygwin-windows10_msvc2013-debug-64bits\test\polyhedron_demo\scene_callback_signaler_autogen\EWIEGA46WW/moc_Callback_signaler.cpp(134) : warning C4273: 'Callback_signaler::progressChanged' : inconsistent dll linkage
        c:\cgal\test\cgal-4.13-ic-42\cmake\platforms\x64_cygwin-windows10_msvc2013-debug-64bits\test\polyhedron_demo\scene_callback_signaler_autogen\ewiega46ww\../../Callback_signaler.h(35) : see previous definition of 'progressChanged'
NMAKE : fatal error U1077: 'C:\PROGRA~2\MICROS~2.0\VC\bin\amd64\cl.exe' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
```
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.13-Ic-42/Polyhedron_Demo/TestReport_afabri_x64_Cygwin-Windows10_MSVC2013-Debug-64bits.gz

The reason is pretty obvious: when the library `scene_callback_signaler` is compiled, CMake automatically adds the definition `-Dcene_callback_signaler_EXPORTS=1`, but in `Callback_signaler.h` the test is on the wrong macro name:
https://github.com/CGAL/cgal/blob/96dcfc070dbf2f8c33c5f86c557f144eed791559/Polyhedron/demo/Polyhedron/Callback_signaler.h#L7-L11

```

**Comment by lrineau** [comment]
The `callback_example.cpp`, on Windows, seems to last forever, and that kills the test suite of the package:
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.13-Ic-42.shtml#Point_set_processing_3_Examples

**Comment by sgiraudot** [comment]
The problem is still here… I'll have to check what is going directly on a Windows machine.

File: Point_set_processing_3/examples/Point_set_processing_3/callback_example.cpp [hunk 1]
```diff
@@ -17,7 +17,8 @@ typedef CGAL::Random_points_on_sphere_3<Point> Generator;
 
 // Concurrency
 #ifdef CGAL_LINKED_WITH_TBB
-typedef CGAL::Parallel_tag Concurrency_tag;
+//typedef CGAL::Parallel_tag Concurrency_tag;
+typedef CGAL::Sequential_tag Concurrency_tag;
```
[the remark this answers is not in the document]
Comment (sgiraudot) [reply Point_set_processing_3/examples/Point_set_processing_3/callback_example.cpp]: Oops, that shouldn't have been committed I just wanted to test the sequential code without changing my configuration… I'll remove that.

**Comment by sgiraudot** [comment]
Alright, the problem should be solved now…

(For reference, [this blog post](https://medium.com/plaxis/embarrassingly-poor-performance-of-regular-point-sets-with-std-unordered-set-418cb413974b) details the problem.)

File: Installation/include/CGAL/config.h [hunk 2]
```diff
@@ -184,7 +184,8 @@
 #if defined(BOOST_NO_0X_HDR_UNORDERED_SET) || \
     defined(BOOST_NO_0X_HDR_UNORDERED_MAP) || \
     defined(BOOST_NO_CXX11_HDR_UNORDERED_SET) || \
-    defined(BOOST_NO_CXX11_HDR_UNORDERED_MAP)
+    defined(BOOST_NO_CXX11_HDR_UNORDERED_MAP) || \
+   (defined(_MSC_VER) && (_MSC_VER == 1800)) // std::unordered_set is very bad in MSVC2013
```
Comment (lrineau) [inline Installation/include/CGAL/config.h]: Either the commit message or that line should have a comment pointing to the web page you saw.

Comment (sgiraudot) [reply Installation/include/CGAL/config.h]: Ok, I amended the commit.

**Pull request merged**